src/com/facebook/buck/go/GoDescriptors.java

```diff
@@ -252,40 +223,125 @@ private static Path extractTestMainGenerator() {
     }
   }
 
-  static Tool getTestMainGenerator(
-      GoBuckConfig goBuckConfig,
-      BuildRuleParams sourceParams,
-      BuildRuleResolver resolver,
-      CxxPlatform cxxPlatform,
-      ProjectFilesystem projectFilesystem) {
-    Optional<Tool> configTool = goBuckConfig.getGoTestMainGenerator(resolver);
-    if (configTool.isPresent()) {
-      return configTool.get();
+  private static BuildTarget createSymlinkTreeTarget(BuildTarget source) {
+    return BuildTarget.builder(source)
+        .addFlavors(ImmutableFlavor.of("symlink-tree"))
+        .build();
+  }
+
+  private static BuildTarget createTransitiveSymlinkTreeTarget(BuildTarget source) {
+    return BuildTarget.builder(source)
+        .addFlavors(ImmutableFlavor.of("transitive-symlink-tree"))
+        .build();
+  }
+
+  private static GoLinkable requireGoLinkable(
+      BuildTarget sourceRule, BuildRuleResolver resolver, GoPlatform platform, BuildTarget target)
+      throws NoSuchBuildTargetException {
+    Optional<GoLinkable> linkable = resolver.requireMetadata(
+        BuildTarget.builder(target)
+            .addFlavors(platform.getFlavor())
+            .build(), GoLinkable.class);
+    if (!linkable.isPresent()) {
+      throw new HumanReadableException(
+          "%s (needed for %s) is not an instance of go_library!",
+          target.getFullyQualifiedName(),
+          sourceRule.getFullyQualifiedName());
     }
+    return linkable.get();
+  }
 
-    // TODO(mikekap): Make a single test main gen, rather than one per test. The generator itself
-    // doesn't vary per test.
+  private static ImmutableSet<GoLinkable> requireGoLinkables(
+      final BuildTarget sourceTarget, final BuildRuleResolver resolver, final GoPlatform platform,
+      ImmutableSet<BuildRule> rules)
+      throws NoSuchBuildTargetException {
+    return FluentIterable.from(rules).transform(new Function<BuildRule, GoLinkable>() {
+      @Override
+      public GoLinkable apply(BuildRule input) {
+        try {
+          return requireGoLinkable(sourceTarget, resolver, platform, input.getBuildTarget());
+        } catch (NoSuchBuildTargetException e) {
+          throw Throwables.propagate(e);
+        }
+      }
+    }).toSet();
+  }
 
-    BuildRuleParams params = sourceParams.copyWithChanges(
-        BuildTarget.builder(sourceParams.getBuildTarget())
-            .addFlavors(ImmutableFlavor.of("make-test-main-gen"))
-            .build(),
-        Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of()),
-        Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of())
-    );
+  private static ImmutableSet<GoLinkable> requireTransitiveGoLinkables(
+      final BuildTarget sourceTarget, final BuildRuleResolver resolver, final GoPlatform platform,
+      ImmutableSet<BuildRule> rules)
+      throws NoSuchBuildTargetException {
+    final ImmutableSet.Builder<GoLinkable> linkables = ImmutableSet.builder();
+    new AbstractBreadthFirstTraversal<BuildRule>(rules) {
```

Comment:
Can you replace this traversal with `requireMetadata` as well? You might need to add a recursive component where every description returns the transitive set of `GoLinkable` that target depends on.